lib/src/tree.rs

```diff
@@ -258,30 +263,31 @@ impl Iterator for TreeEntriesIterator<'_> {
     type Item = (RepoPath, TreeValue);
 
     fn next(&mut self) -> Option<Self::Item> {
-        loop {
-            // First return results from any subdirectory we're currently visiting.
-            if let Some(subdir_iter) = &mut self.subdir_iterator {
-                if let Some(item) = subdir_iter.next() {
-                    return Some(item);
-                }
-                self.subdir_iterator = None;
-            }
-            let entry = self.entry_iterator.next()?;
-            match entry.value() {
-                TreeValue::Tree(id) => {
-                    let subtree = self.tree.known_sub_tree(entry.name(), id);
-                    self.subdir_iterator =
-                        Some(Box::new(TreeEntriesIterator::new(subtree, self.matcher)));
-                }
-                other => {
-                    let path = self.tree.dir().join(entry.name());
-                    if !self.matcher.matches(&path) {
-                        continue;
+        while let Some(top) = self.stack.last_mut() {
+            if let Some(entry) = top.entry_iterator.next() {
+                let path = top.tree.dir().join(entry.name());
+                match entry.value() {
+                    TreeValue::Tree(id) => {
+                        let subtree = top.tree.known_sub_tree(entry.name(), id);
+                        let entry_iterator = subtree.entries_non_recursive();
+                        let entry_iterator: TreeEntriesNonRecursiveIterator<'static> =
+                            unsafe { std::mem::transmute(entry_iterator) };
+                        self.stack.push(TreeEntriesDirItem {
+                            entry_iterator,
+                            tree: Box::pin(subtree),
+                        });
```

Comment:
I would add a helper function that encapsulates the unsafe bits.

`entry_iterator` should be borrowed from the boxed `subtree`, not from the stack.

Reply:
Yeah, that also helps reduce duplication and makes it more readable. And good catch about the borrow. Thanks!